### ISSUE TITLE
神经网络面试问题汇总

### ISSUE THREAD

**Issue by MiaoRain** [MiaoRain/lund#5]
**`大集合`**
https://github.com/MiaoRain/Awesome-algorithm-interview
**Deeper learning 500问** 为主
https://github.com/MiaoRain/DeepLearning-500-questions
**interview notes**
https://github.com/MiaoRain/Interview-Notes
一次卷积的计算量到底有多少 ?
https://blog.csdn.net/mtj66/article/details/80509713

**神经网络/深度学习本质是分类，那么目标识别是整张图片的分类，目标检测问题是anchor级别的分类，图像分割是像素级别的分类**

![image](https://user-images.githubusercontent.com/47746159/76629768-1baeb180-653f-11ea-8fe0-fd0f856e2344.png)

![image](https://user-images.githubusercontent.com/47746159/76629829-3d0f9d80-653f-11ea-9742-181fcb0d1535.png)

![image](https://user-images.githubusercontent.com/47746159/76631184-8c56cd80-6541-11ea-8133-713490bd0d08.png)

![image](https://user-images.githubusercontent.com/47746159/76631694-5fef8100-6542-11ea-8567-729d95cbf48f.png)
欠拟合和过拟合问题
https://zhuanlan.zhihu.com/p/72038532
https://kaiyuanzhang.com/2018/05/02/overfitting-and-underfitting/

卷积核一般都把size设为**奇数**，主要有以下两个原因：
保护位置信息：保证了 **锚点** 刚好在中间，方便以模块中心为标准进行滑动卷积，避免了位置信息发生 偏移 。
padding时对称：保证了 **padding** 时，图像的两边依然相 对称 。

**为什么pooling层？**
本质上，是在精简feature map数据量的同时，最大化保留空间信息和特征信息. 如果一个象限中检测到一个特征值，那么最大值就可以作为代表。如果没有检测到特征，那么数值本身不会很大
做窗口滑动卷积的时候，卷积值就代表了整个窗口的特征。因为滑动的窗口间有大量重叠区域，出来的卷积值有冗余，进行最大pooling或者平均pooling就是减少冗余。
(1) 增大感受野 or 降维
(2) 增加平移不变性。我们希望目标的些许位置的移动，能得到相同的结果。因为pooling不断地抽象了区域的特征而不关心位置，所以pooling一定程度上增加了平移不变性。
(3) 使网络更容易优化，pooling是每个featuremap单独做降采样，与基于卷积的降采样相比，不需要参数，更容易优化。
(4)、减小计算量
计算量与通道数、特征图长和特征图宽的乘积呈正比，若输入通道数、特征图宽和特征图高分别为c、w、h，则经过stride为2的pooling层，输出为c、w/2、h/2，计算量减小为1/4。

why convolution layer？

![image](https://user-images.githubusercontent.com/47746159/77231600-64e4ae00-6b9c-11ea-87e7-ae17283be1ac.png)
卷积有很好的平移不变性，但是maxpool会破坏此特性。
![image](https://user-images.githubusercontent.com/47746159/77231902-1afcc780-6b9e-11ea-9dc0-7563fd925712.png)

Why residual net work？

**Comment by MiaoRain** [comment]
![image](https://user-images.githubusercontent.com/47746159/77226206-d873c480-6b76-11ea-93c5-df3a5d178487.png)

**Comment by MiaoRain** [comment]
![image](https://user-images.githubusercontent.com/47746159/77226299-8d0de600-6b77-11ea-95bf-cddc5dbb794c.png)

**Comment by MiaoRain** [comment]
![image](https://user-images.githubusercontent.com/47746159/77226400-6603e400-6b78-11ea-8460-b2ebef74d180.png)

**Comment by MiaoRain** [comment]
![image](https://user-images.githubusercontent.com/47746159/77226487-06f29f00-6b79-11ea-90ad-5b108efc9e40.png)

**Comment by MiaoRain** [comment]
![image](https://user-images.githubusercontent.com/47746159/77226523-5638cf80-6b79-11ea-8d14-a7b0b66e0a21.png)

**Comment by MiaoRain** [comment]
1*1卷积的作用
1*1的卷积是多个feature channels之间的线性叠加，并且可以调节channels的大小
1、降维（ dimension reductionality ）。比如，一张500 * 500且厚度depth为100 的图片在20个filter上做1*1的卷积，那么结果的大小为500*500*20。2、加入非线性。卷积层之后经过激励层，1*1的卷积在前一层的学习表示上添加了非线性激励（ non-linear activation ），提升网络的表达能力；

**Comment by MiaoRain** [comment]
**为什么全局信息和局部信息(像素级)很难同时得到？**
从感受野角度考虑，感受野较小时，获取局部信息能力强，反之，全局信息较强。
**目标检测和感受野的总结和想法**
https://zhuanlan.zhihu.com/p/108493730

**Comment by MiaoRain** [comment]
**不同优化器的选择**
https://www.cnblogs.com/guoyaohua/p/8542554.html

**Comment by MiaoRain** [comment]
**Padding 作用**
![image](https://user-images.githubusercontent.com/47746159/77358303-e594dd00-6d49-11ea-8978-d0225c6baf10.png)

**Comment by MiaoRain** [comment]
**上下采样**
https://zhuanlan.zhihu.com/p/87434515

**Comment by MiaoRain** [comment]
**卷积神经网络参数量和计算量的计算**
https://blog.csdn.net/liuweiyuxiang/article/details/99637649

**Comment by MiaoRain** [comment]
技术不单单是掌握的技术栈，还有对目前主流技术，发展趋势等，的看法与理解。个人认为能合理的审视当前的形势，比实际掌握某个技术更重要

**Comment by MiaoRain** [comment]
**深度学习中的经典基础网络结构（backbone）总结**
https://blog.csdn.net/kuweicai/article/details/102789420?depth_1-utm_source=distribute.pc_relevant.none-task-blog-BlogCommendFromBaidu-5&utm_source=distribute.pc_relevant.none-task-blog-BlogCommendFromBaidu-5
![image](https://user-images.githubusercontent.com/47746159/79266974-ad8c4000-7e98-11ea-9fba-01758ba387ed.png)

**Comment by MiaoRain** [comment]
**显存占用 = 模型显存占用 + batch_size × 每个样本的显存占用**
https://zhuanlan.zhihu.com/p/31558973

**Comment by MiaoRain** [comment]
**梯度爆炸和消失**



**1. BN层作用**
**深度学习（二十九）Batch Normalization 学习笔记**
https://blog.csdn.net/hjimce/article/details/50866313
**深度学习中的Normalization模型**
https://zhuanlan.zhihu.com/p/43200897

关于Normalization的有效性，有以下几个主要观点：(1) 主流观点，Batch Normalization调整了数据的分布，不考虑激活函数，它让每一层的输出归一化到了均值为0方差为1的分布，这保证了梯度的有效性，目前大部分资料都这样解释，比如BN的原始论文认为的缓解了Internal Covariate Shift(ICS)问题。(2) 可以使用更大的学习率，文[2]指出BN有效是因为用上BN层之后可以使用更大的学习率，从而跳出不好的局部极值，增强泛化能力，在它们的研究中做了大量的实验来验证。(3) 损失平面平滑。文[3]的研究提出，BN有效的根本原因不在于调整了分布，因为即使是在BN层后模拟ICS，也仍然可以取得好的结果。它们指出，BN有效的根本原因是平滑了损失平面。之前我们说过，Z-score标准化对于包括孤立点的分布可以进行更平滑的调整。
链接：https://www.zhihu.com/question/38102762/answer/607815171

![image](https://user-images.githubusercontent.com/47746159/81508139-952a0c80-9302-11ea-9124-3359917c981d.png)

**Comment by MiaoRain** [comment]
**深度学习最常用的算法:Adam优化算法**
https://juejin.im/entry/5983115f6fb9a03c50227fd4

**Comment by MiaoRain** [comment]
**Metrics**
IOU和dice coefficient
![image](https://user-images.githubusercontent.com/47746159/81507046-c272bc80-92fa-11ea-94d5-b0e4069425a7.png)

- 目标检测评价指标：mAP、Precision、Recall、AP、IOU

https://blog.csdn.net/Gentleman_Qin/article/details/84519388?utm_medium=distribute.pc_relevant.none-task-blog-BlogCommendFromMachineLearnPai2-1.nonecase&depth_1-utm_source=distribute.pc_relevant.none-task-blog-BlogCommendFromMachineLearnPai2-1.nonecase


![image](https://user-images.githubusercontent.com/47746159/81549851-ae24d300-937f-11ea-88cb-6477a0092f13.png)

**Comment by MiaoRain** [comment]
**[矩阵求导]神经网络反向传播梯度计算数学原理**
https://zhuanlan.zhihu.com/p/32368246

**Comment by MiaoRain** [comment]
**NMS非极大值抑制** 
非极大值抑制的流程如下：

根据置信度得分进行排序
选择置信度最高的比边界框添加到最终输出列表中，将其从边界框列表中删除
计算所有边界框的面积
计算置信度最高的边界框与其它候选框的IoU。
删除IoU大于阈值的边界框
重复上述过程，直至边界框列表为空。
https://zhuanlan.zhihu.com/p/37489043

**Comment by MiaoRain** [comment]
**Tensorflow与pytorch的异同点**
Tensorflow的设计理念称之为计算流图，在编写程序时，首先构筑整个系统的graph，代码并不会直接生效，这一点和python的其他数值计算库（如Numpy等）不同，graph为静态的，类似于docker中的镜像。然后，在实际的运行时，启动一个session，程序才会真正的运行。这样做的好处就是：避免反复地切换底层程序实际运行的上下文，tensorflow帮你优化整个系统的代码。我们知道，很多python程序的底层为C语言或者其他语言，执行一行脚本，就要切换一次，是有成本的，tensorflow通过计算流图的方式，帮你优化整个session需要执行的代码，还是很有优势的。
————————————————
版权声明：本文为CSDN博主「于小勇」的原创文章，遵循CC 4.0 BY-SA版权协议，转载请附上原文出处链接及本声明。
原文链接：https://blog.csdn.net/weixin_36670529/article/details/94595043

**PyTorch和TensorFlow哪家强：九项对比读懂各自长项短板**
https://www.jiqizhixin.com/articles/2017-08-31-3

**Comment by MiaoRain** [comment]
**pytorch Conv 卷积层源码**
https://blog.csdn.net/moshiyaofei/article/details/102532322

**Comment by MiaoRain** [comment]
**池化层（pooling）的反向传播是怎么实现的**
https://blog.csdn.net/Jason_yyz/article/details/80003271

**Comment by MiaoRain** [comment]
**深度神经网络中的感受野(Receptive Field)**
https://zhuanlan.zhihu.com/p/28492837

**Comment by MiaoRain** [comment]
**dilated convolution (扩张卷积) 输入输出特征图尺寸大小关系以及 receptive field (感受野) 计算**
https://blog.csdn.net/xiaojiajia007/article/details/80810813

**Comment by MiaoRain** [comment]
**07 卷积神经网络反向传播原理**
https://www.youtube.com/watch?v=4_8tovnXT7w

**Comment by MiaoRain** [comment]
**从 SGD 到 Adam —— 6大常见优化算法总结**
https://zhuanlan.zhihu.com/p/64113429
https://zhuanlan.zhihu.com/p/22252270

**Comment by MiaoRain** [comment]
**Batch Normalization原理与实战**
https://zhuanlan.zhihu.com/p/34879333

**Comment by MiaoRain** [comment]
**理解 Deformable Convolutional Networks**
https://zhuanlan.zhihu.com/p/52476083

**Comment by MiaoRain** [comment]
图像语义分割准确率度量方法总结😊
https://zhuanlan.zhihu.com/p/38236530

**Comment by MiaoRain** [comment]
**深度学习 | Dropout对深度学习的影响**
https://zhuanlan.zhihu.com/p/96536112

**Comment by MiaoRain** [comment]
**pooling有什么用呢？或者说为什么需要pooling呢？**

原因有几个：

(1) 增大感受野

所谓感受野，即一个像素对应回原图的区域大小，假如没有pooling，一个3*3，步长为1的卷积，那么输出的一个像素的感受野就是3*3的区域，再加一个stride=1的3*3卷积，则感受野为5*5，我们看左上角像素的传播就明白了。

依次，再多一个卷积，则为7*7，如果想看到224*224的全图，大家可以算算需要多少卷积层。



假如我们在每一个卷积中间加上3*3的pooling呢？很明显感受野迅速增大，这就是pooling的一大用处。感受野的增加对于模型的能力的提升是必要的，正所谓“一叶障目则不见泰山也”。

(2) 平移不变性

我们希望目标的些许位置的移动，能得到相同的结果。因为pooling不断地抽象了区域的特征而不关心位置，所以pooling一定程度上增加了平移不变性。

(3) 容易优化，pooling是每个featuremap单独做降采样，与基于卷积的降采样相比，不需要参数，更容易优化。
